returnn/dataset.py

```diff
@@ -0,0 +1,90 @@
+__all__ = ["ExtractDatasetStatisticsJob"]
+
+from sisyphus import *
+
+import numpy
+import os
+import shutil
+import subprocess
+
+from i6_core.returnn.config import ReturnnConfig
+from i6_core.util import create_executable
+
+
+class ExtractDatasetStatisticsJob(Job):
+    """
+    Runs the RETURNN tool dump-dataset with statistic extraction.
+    Collects mean and std-var for each feature as file and in total as sisyphus var.
+
+    Outputs:
+
+    Variable out_mean: a global mean over all sequences and features
+    Variable out_std_dev: a global std-dev over all sequences and features
+
+    Path out_mean_file: a text file with #feature entries for the mean
+    Path out_std_dev_file: a text file with #features entries for the standard deviation
+    """
+
+    def __init__(self, returnn_config, returnn_python_exe=None, returnn_root=None):
+        """
+
+        :param ReturnnConfig returnn_config:
+        :param Path|None returnn_python_exe:
+        :param Path|None returnn_root:
+        """
+
+        self.returnn_config = returnn_config
+        self.returnn_python_exe = (
+            returnn_python_exe
+            if returnn_python_exe is not None
+            else gs.RETURNN_PYTHON_EXE
+        )
+        self.returnn_root = (
+            returnn_root if returnn_root is not None else gs.RETURNN_ROOT
+        )
+        self.out_mean = self.output_var("mean_var")
+        self.out_std_dev = self.output_var("std_dev_var")
+
+        self.out_mean_file = self.output_path("mean")
+        self.out_std_dev_file = self.output_path("std_dev")
+
+        self.rqmt = {"cpu": 2, "mem": 4, "time": 8}
+
+    def tasks(self):
+        yield Task("run", rqmt=self.rqmt)
+
+    def run(self):
+        self.returnn_config.write("returnn.config")
+
+        command = [
+            self.returnn_python_exe.get(),
+            os.path.join(self.returnn_root.get(), "tools/dump-dataset.py"),
+            "returnn.config",
+            "--endseq -1",
+            "--stats",
+            "--dump_stats stats",
+        ]
+
+        create_executable("rnn.sh", command)
+        subprocess.check_call(["./rnn.sh"])
+
+        shutil.copy("stats.mean.txt", self.out_mean_file.get_path())
+        shutil.copy("stats.std_dev.txt", self.out_std_dev_file.get_path())
+
+        total_mean = 0
+        total_var = 0
+
+        mean_file = open("stats.mean.txt")
+        std_dev_file = open("stats.std_dev.txt")
+
+        # compute the total mean and std-dev in an iterative way
+        for i, (mean, std_dev) in enumerate(zip(mean_file, std_dev_file)):
```

Comment:
Well... this part gave me a lot of headaches in the past. The problem is the RETURNN tool only computes mean and variance per feature axis, but for TTS I needed it globally. So I did some research how I can transform multiple means and variances into a single mean and variance. I think I should add a test case for this to confirm it is correct. I tested it locally, but this might be better be somewhere explicit.